Kafka Stream Publisher  may throw `ArrayIndexOutOfBoundsException`

Using `KafkaTools.consumeToTable` (with a little help of setting `StreamPublisherBase.CHUNK_SIZE` to `32`; but happens naturally by altering Kafka's `max.poll.records`) we can achieve an `ArrayIndexOutOfBoundsException`.

Occasionally, upon entering `consumeRecords` it is possible for 1) `this.chunks` to be non-null (so we continue to use the existing chunks), and 2) `remaining` (calculated via `chunks[0].capacity() - chunks[0].size();`) is `0`.

As we consume records we pre-decrement the remaining from 0 to -1, which causes the AIOBE.
```java
            for (ConsumerRecord<?, ?> record : records) {
                if (--remaining == 0) {
```

The total failure looks similar to this:
```
ph-updateExecutor-11 | .AsyncClientErrorNotifier | Error in table update: Multiple errors encountered while ingesting stream: 
Cause 0: java.lang.ArrayIndexOutOfBoundsException: Index 32 out of bounds for length 32
Cause 1: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 2: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 3: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 4: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 5: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 6: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 7: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 8: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 9: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 10: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 11: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 12: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 13: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 14: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 15: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 16: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 17: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 18: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32
Cause 19: java.lang.IllegalStateException: Publisher chunks have size mismatch: 33, 32, 32, 32, 32, 32, 32, 32, 32, 32

	at io.deephaven.util.MultiException.maybeWrapInMultiException(MultiException.java:55)
	at io.deephaven.stream.StreamToBlinkTableAdapter.deliverFailures(StreamToBlinkTableAdapter.java:330)
	at io.deephaven.stream.StreamToBlinkTableAdapter.run(StreamToBlinkTableAdapter.java:308)
	at io.deephaven.engine.updategraph.impl.PeriodicUpdateGraph$UpdateSourceRefreshNotification.run(PeriodicUpdateGraph.java:1879)
	at io.deephaven.engine.updategraph.impl.PeriodicUpdateGraph.runNotification(PeriodicUpdateGraph.java:1322)
	at io.deephaven.engine.updategraph.impl.PeriodicUpdateGraph$ConcurrentNotificationProcessor.processSatisfiedNotifications(PeriodicUpdateGraph.java:1373)
	at io.deephaven.engine.updategraph.impl.PeriodicUpdateGraph$NotificationProcessorThreadFactory.lambda$newThread$0(PeriodicUpdateGraph.java:1911)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 32 out of bounds for length 32
	at io.deephaven.chunk.WritableIntChunk.add(WritableIntChunk.java:79)
	at io.deephaven.kafka.ingest.KafkaStreamPublisher.consumeRecords(KafkaStreamPublisher.java:248)
	at io.deephaven.kafka.KafkaTools$SimpleKafkaRecordConsumer.consume(KafkaTools.java:1636)
	at io.deephaven.kafka.ingest.KafkaIngester.pollOnce(KafkaIngester.java:383)
	at io.deephaven.kafka.ingest.KafkaIngester.consumerLoop(KafkaIngester.java:301)
	... 1 more
Cause 1:
java.lang.ArrayIndexOutOfBoundsException: Index 32 out of bounds for length 32
	at io.deephaven.chunk.WritableIntChunk.add(WritableIntChunk.java:79)
	at io.deephaven.kafka.ingest.KafkaStreamPublisher.consumeRecords(KafkaStreamPublisher.java:248)
	at io.deephaven.kafka.KafkaTools$SimpleKafkaRecordConsumer.consume(KafkaTools.java:1636)
	at io.deephaven.kafka.ingest.KafkaIngester.pollOnce(KafkaIngester.java:383)
	at io.deephaven.kafka.ingest.KafkaIngester.consumerLoop(KafkaIngester.java:301)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

--- a/extensions/kafka/src/main/java/io/deephaven/kafka/ingest/KafkaStreamPublisher.java
+++ b/extensions/kafka/src/main/java/io/deephaven/kafka/ingest/KafkaStreamPublisher.java
@@ -203,7 +203,8 @@ public class KafkaStreamPublisher extends StreamPublisherBase implements Consume
                     : null;
 
             for (ConsumerRecord<?, ?> record : records) {
-                if (--remaining == 0) {
+                // Note we only flush if we are about to overflow the chunks.
+                if (--remaining < 0) {
                     if (keyChunk != null) {
                         flushKeyChunk(keyChunk, chunks);
                     }
@@ -217,8 +218,9 @@ public class KafkaStreamPublisher extends StreamPublisherBase implements Consume
                     chunks = getChunksToFill();
                     checkChunkSizes(chunks);
 
-                    remaining = chunks[0].capacity() - chunks[0].size();
-                    Assert.gtZero(remaining, "remaining");
+                    // Note that remaining should account for the row we are about to write.
+                    remaining = chunks[0].capacity() - chunks[0].size() - 1;
+                    Assert.geqZero(remaining, "remaining");
 
                     if (kafkaPartitionColumnIndex >= 0) {
                         partitionChunk = chunks[kafkaPartitionColumnIndex].asWritableIntChunk();

--- a/extensions/kafka/src/main/java/io/deephaven/kafka/ingest/KafkaStreamPublisher.java
+++ b/extensions/kafka/src/main/java/io/deephaven/kafka/ingest/KafkaStreamPublisher.java
@@ -203,8 +203,7 @@ public class KafkaStreamPublisher extends StreamPublisherBase implements Consume
                     : null;
 
             for (ConsumerRecord<?, ?> record : records) {
-                // Note we only flush if we are about to overflow the chunks.
-                if (--remaining < 0) {
+                if (remaining == 0) {
                     if (keyChunk != null) {
                         flushKeyChunk(keyChunk, chunks);
                     }
@@ -218,9 +217,8 @@ public class KafkaStreamPublisher extends StreamPublisherBase implements Consume
                     chunks = getChunksToFill();
                     checkChunkSizes(chunks);
 
-                    // Note that remaining should account for the row we are about to write.
-                    remaining = chunks[0].capacity() - chunks[0].size() - 1;
-                    Assert.geqZero(remaining, "remaining");
+                    remaining = chunks[0].capacity() - chunks[0].size();
+                    Assert.gtZero(remaining, "remaining");
 
                     if (kafkaPartitionColumnIndex >= 0) {
                         partitionChunk = chunks[kafkaPartitionColumnIndex].asWritableIntChunk();
@@ -275,6 +273,8 @@ public class KafkaStreamPublisher extends StreamPublisherBase implements Consume
                         bytesProcessed += valueBytes;
                     }
                 }
+
+                --remaining;
             }
             if (keyChunk != null) {
                 flushKeyChunk(keyChunk, chunks);

